cpuid/src/commons.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use kvm::CpuId;
+use kvm_bindings::kvm_cpuid_entry2;
+
+const INTEL: &str = "GenuineIntel";
+const AMD: &str = "AuthenticAMD";
+
+extern "C" {
+    fn c_get_cpuid(
+        opcode: u32,
+        count: u32,
+        eax: *mut u32,
+        ebx: *mut u32,
+        ecx: *mut u32,
+        edx: *mut u32,
+    ) -> i32;
+}
+
+fn get_cpuid(opcode: u32, count: u32) -> Option<kvm_cpuid_entry2> {
+    unsafe {
+        let mut eax: u32 = 0;
+        let mut ebx: u32 = 0;
+        let mut ecx: u32 = 0;
+        let mut edx: u32 = 0;
+
+        if c_get_cpuid(opcode, count, &mut eax, &mut ebx, &mut ecx, &mut edx) != 1 {
+            return None;
+        }
+
+        Some(kvm_cpuid_entry2 {
+            function: opcode,
+            index: count,
+            flags: 0,
+            eax: eax,
+            ebx: ebx,
+            ecx: ecx,
+            edx: edx,
+            padding: [0, 0, 0],
+        })
+    }
+}
+
+/// Extracts the CPU vendor id from leaf 0x0.
+///
+pub fn get_vendor_id() -> String {
+    if let Some(vendor_entry) = get_cpuid(0, 0) {
+        let bytes: [u8; 12] =
+            unsafe { std::mem::transmute([vendor_entry.ebx, vendor_entry.edx, vendor_entry.ecx]) };
+        return std::str::from_utf8(&bytes).unwrap().to_string();
+    }
+
+    "".to_string()
+}
+
+/// Adds to the `cpuid` all the host leaves corresponding to the `opcode`.
+///
+pub fn extend_cpuid(cpuid: &mut CpuId, opcode: u32) {
+    let mut entries = cpuid.mut_entries_slice().to_vec();
```

Comment:
nit-ish: There seems to be some unnecessary copying-around of data here. I.e. For each leaf (opcode) you want to add, the entire list of entries is copied (`.to_vec()`), then filtered (`.retain()` - possibly some more copying here), then copied again (`.set_entries()`).

Reply:
I removed the call to `retain()`